### DISCUSSION TITLE
Possibility to add scopes to resolvers

### DISCUSSION THREAD

**Discussion by dberardo-com** [TriPSs/nestjs-query#316] [Q&A]
I would like to pre-filter (scope) records in the cRUD resolvers before all other filters , aggregates, etc. gets applied.

Is this possible out of the box?

In particular I have e.g. some records in the DB and I would like the current user to only be able to see their own records (e.g. record.ownerId == userId) before applying filtering.

I know the concept of guards but this is not what I am looking for. I want to first reduce the search base of records for the service and then apply the transformations.

Same goes for updates deletes etc

**Comment by TriPSs** [comment]
I think what you are looking for is the [authorize decorator](https://tripss.github.io/nestjs-query/docs/graphql/authorization/#authorize-decorator)

**Reply by dberardo-com** [reply]
I think I can give that a try, I have got the bed to join multiple tables in my authorization logic, so perhaps my filter will simply explode, as I would need to insert a list of IDs of enabled resources I suppose.

Currently I was creating a "scope" meaning I was using the typeorm query builder too prepare a query without materializing it and then passing it to the downstream resolvers. Here an example:


Resolver.dashboards --> this.dashboardService.scopeOn(context.user).find({})


this.dashboardService.scopeOn(user: any)  --> this.userRepo.createQueryBuilder(...).leftJoin(... usergroups ...).leftJoin(...other stuff...).where(....).addSelect(...)

Is it possible to use the authorize method without having to materialize the query first, i.e. can one return only the prorated query builder?

**Reply by dberardo-com** [reply]
To be more specific, I am looking for something that can make me flexible model relationships such as: users have read access to all dashboards they own, or that have been defined in a dashboardGroup which is linked (many-to-many) to a userGroup that condoms the user via a read/write permission.

User -[ belongsto ]-> UserGroup -[ canread ]-> DashboardGroup <-[ belongsto ]- Dashboard

This is to give an example. I am not sure if this can be modeled via a filter within the framework and this I would rather rely on passing the query builder downstream to the resolvers like I am doing currently
